### PR TITLE
Rdkit images API

### LINKED ISSUE #740: Molecule API
## Summary

We need an API handling molecules.
- [ ] image generation
- [ ] substructure & SMARTS search
- [ ] substructure image highlighting 
- [ ] maximum common substructure

For reference, use ChEmbVis: https://github.com/puehringer/ChEmbVis/blob/main/api/api/molecule.py

### PR DESCRIPTION
**Pull request by dvmarkusvogl**
Shouldn't change existing behaviour if you don't have rdkit installed.

If you have rdkit / rdkit-pypi installed, this will add an api with the prefix /api/images that converts molecules into smiles.

For more info about behaviour look at tests.

Also the API is still open for discussion.

Not sure if we need the /substructures endpoint

Suggested reference impl:
https://github.com/puehringer/ChEmbVis/blob/main/api/api/molecule.py

Closes #740

### CODE REVIEW

File: tdp_core/mol_img/img_api.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+def draw_maximum_common_substructure_molecule(structures: List[SmilesMolecule]):
+    unique = [m.mol for m in set(structures)]
+    mcs = maximum_common_substructure_query_mol(unique)
+    print(mcs)
+    return draw(mcs)
+
+
+@app.post("/substructures/{substructure}")
+def substructures_count(structures: Set[SmilesMolecule], substructure: SmilesSmartsMolecule) -> SubstructuresResponse:
+    """TODO: maybe put to another file"""
```
Comment (dvchristianbors) [inline tdp_core/mol_img/img_api.py:45]: Do you mean because it's a substructure? For the moment I would leave it here.

**Review by dvchristianbors — DISMISSED** [review]
I manually tested the endpoint as well as ran tests locally and in the docker container. Functionally, the PR does what it promises :+1:

File: setup.py [hunk 2]
```diff
@@ -49,9 +49,7 @@ def requirements(file):
     # requirements files see:
     # https://packaging.python.org/en/latest/requirements.html
     install_requires=requirements("requirements.txt"),
-    extras_require={
-        "develop": requirements("requirements_dev.txt"),
-    },
+    extras_require={"develop": requirements("requirements_dev.txt"), "rdkit": ["rdkit-pypi"]},
```
[the remark this answers is not in the document]
Comment (dvmarkusvogl) [reply setup.py:52]: @thinkh : works as intended / specified. If you don't have rdkit, you don't get the rdkit-endpoints. If you install tdp-core[rdkit] (it's called an extras / a package flavour in other languages), you get it.

Imo our project structure doesn't perfectly support the concept of optional packages and I think it could be a nice standalone tdp package. (However: Complexity of yet another package)
Or we just unconditionally take it into tdp_core, it's just 20mb of mostly binaries after all.

[the remark this answers is not in the document]
Comment (thinkh) [reply setup.py:52]: Ok, thanks for the clarification. Do you know how I install `tdp-core[rdkit]` as part of the `docker-compose build api` (or `docker-compose up`)? We would need to call this somehow later when building the Docker images for the deployment.

[the remark this answers is not in the document]
Comment (dvmarkusvogl) [reply setup.py:52]: Assuming the generator isn't messing it up, just add
replace `-e /phovea/tdp_core` by `-e /phovea/tdp_core[rdkit]` in the requirements_workspaces.txt or to the product like this:
https://github.com/datavisyn/ordino_reprovisyn_product/pull/13/files (testing that currently in progress)

[the remark this answers is not in the document]
Comment (dvmarkusvogl) [reply setup.py:52]: Checkout/build seems to work. But I'm having some port conflict so I can't docker-compose-up.

Regarding the default-title:
To my understanding it should be a tag, right? I added that.

[the remark this answers is not in the document]
Comment (puehringer) [reply setup.py:52]: As I said, I think we should remove the flavor (as it isn't really supported by our infra right now) and just add it to the dependencies. In the end, it doesn't matter if the server image is 20MB more or less. 
@dvmarkusvogl could you do this? Then you also don't need to check for the availability of the module. 
@thinkh  do you agree?

[the remark this answers is not in the document]
Comment (thinkh) [reply setup.py:52]: Yes, I agree.

[the remark this answers is not in the document]
Comment (dvmarkusvogl) [reply setup.py]: Ok, i made rdkit mandatory (by now you don't even need rdkit-pypi, the main rdkit is now the same, suggested package)

I also deleted some code for handling a lack of rdkit

File: tdp_core/server/main.py [hunk 3]
```diff
@@ -1,3 +1,6 @@
+from fastapi import FastAPI
+
 from .visyn_server import create_visyn_server
 
+app = FastAPI()  # needed for pycharm to detect this main
```
Comment (puehringer) [inline tdp_core/server/main.py:5]: We should probably remove this now.

File: tdp_core/mol_img/util/molecule.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    mcs = maximum_common_substructure_query_mol([structure, align])
+    if mcs:
+        TemplateAlign.rdDepictor.Compute2DCoords(mcs)
+        TemplateAlign.AlignMolToTemplate2D(structure, mcs, clearConfs=True)
+        # Enable this to show substructore highlights of alignment
+        # return mcs
+
+
+def murcko(structure):
+    return MurckoScaffold.GetScaffoldForMol(structure)
```
Comment (puehringer) [inline tdp_core/mol_img/util/molecule.py:25]: I think we can remove this utility function as it is only used once, and it is easier to understand what is happening if no wrapping occurs.

File: tdp_core/mol_img/util/molecule.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+def aligned(structure, align):
+    if not align:
+        return None
+    TemplateAlign.rdDepictor.Compute2DCoords(structure)
+    mcs = maximum_common_substructure_query_mol([structure, align])
+    if mcs:
+        TemplateAlign.rdDepictor.Compute2DCoords(mcs)
+        TemplateAlign.AlignMolToTemplate2D(structure, mcs, clearConfs=True)
+        # Enable this to show substructore highlights of alignment
+        # return mcs
```
Comment (puehringer) [inline tdp_core/mol_img/util/molecule.py:21]: I suppose the `AlignMolToTemplate2D` modifies the input, such that returning anything is not required?

File: tdp_core/server/visyn_server.py [hunk 6]
```diff
@@ -121,7 +121,8 @@ def create_visyn_server(
     # Load all namespace plugins as WSGIMiddleware plugins
     for p in router_plugins:
         _log.info(f"Registering router: {p.id}")
-        app.include_router(p.load().factory())
+        prefix = getattr(p, "namespace", "")  # consistent behaviour with flask plugins
+        app.include_router(p.load().factory(), prefix=prefix)
```
Comment (puehringer) [inline tdp_core/server/visyn_server.py:125]: Let's move away from using the registry for appending routers. 

Could you: 
1) Revert this change (just to keep it in-line with the existing behaviour --> we should look at that separately)
2) Add the `init_app(app: FastAPI)` callback to the VisynPlugin in the `__init__.py` and register the router there (via standard FastAPI machanism). 
That's just much clearer imo.

File: tdp_core/tests/test_rdkit_img.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    res = client.get(f"/api/image/{s}")
+    assert res.status_code == 422
+
+
+@pytest.mark.parametrize("mol, expected", mol_expected.items())
+def test_valid(client, mol, expected):
+    res = client.get(f"/api/image/{mol}")
+    assert res.status_code == 200
+    assert res.headers.get("content-type").startswith("image/svg")
+    hash_compare(res.content, expected)
```
Comment (puehringer) [inline tdp_core/tests/test_rdkit_img.py:35]: Do we really want to hash-compare the SVGs? What if a new version of RDKit just _slightly_ changes a color-coding, rotation, etc.? I know that this is meant as "snapshot" test here, but I think it is sufficient to check if a `<svg` including a `<line` is returned, everything else is probably too much. Image similarity test w/ threshold would be interesting though.

**Review by puehringer — COMMENTED** [review]
Some open things, mainly open discussion points.

File: tdp_core/tests/test_rdkit_img.py (hunk 7 above)
Comment (dvmarkusvogl) [reply tdp_core/tests/test_rdkit_img.py:35]: in many rdkit-error-cases, it returns an SVG with a big ? in it (technically still an `<svg` with a `<line`) and i'd rather take the risk of having to change hashes once than not detecting that after upgrading all of our images are question marks.

> Image similarity test w/ threshold would be interesting though.
that sounds like 10x complexity

**Review by thinkh — CHANGES_REQUESTED** [review]
While implementing the renderer and using the test dataset I found two bugs in the API.

# Structures with `/` do not work

There are structures that contain a `/` (e.g. `C(=C/Cl)\\Cl`) which results in an error 404. URI encoding with the structure to `C(%3DC%2FCl)%5CCl` does not solve the problem. Probably we cannot use the structure as part of the URI and we need to switch to a URL parameter (e.g. `?structure=...`).

grafik

http://localhost:8080/api/image/C(%3DC%2FCl)%5CCl


# MCS does not always work

MCS with the following structures as request body responds with an error 500.

```json
[
    "C1COCCO1",
    "CC(COC)OC",
    "CC1(OCCO1)C",
    "CCCCCCCO",
    "CCCCCCO"
]
```

grafik


**Response**

```json
{"detail":"THIS STACKTRACE IS SHOWN IN DEVELOPMENT MODE ONLY. IN PRODUCTION, ONLY THE SPECIFIC ERROR MESSAGE IS SHOWN!Traceback (most recent call last):\n  File \"/usr/local/lib/python3.10/site-packages/anyio/streams/memory.py\", line 94, in receive\n    return self.receive_nowait()\n  File \"/usr/local/lib/python3.10/site-packages/anyio/streams/memory.py\", line 89, in receive_nowait\n    raise WouldBlock\nanyio.WouldBlock\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.10/site-packages/starlette/middleware/base.py\", line 41, in call_next\n    message = await recv_stream.receive()\n  File \"/usr/local/lib/python3.10/site-packages/anyio/streams/memory.py\", line 114, in receive\n    raise EndOfStream\nanyio.EndOfStream\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/phovea/tdp_core/tdp_core/middleware/exception_handler_middleware.py\", line 14, in dispatch\n    return await call_next(request)\n  File \"/usr/local/lib/python3.10/site-packages/starlette/middleware/base.py\", line 44, in call_next\n    raise app_exc\n  File \"/usr/local/lib/python3.10/site-packages/starlette/middleware/base.py\", line 34, in coro\n    await self.app(scope, request.receive, send_stream.send)\n  File \"/usr/local/lib/python3.10/site-packages/starlette/exceptions.py\", line 82, in __call__\n    raise exc\n  File \"/usr/local/lib/python3.10/site-packages/starlette/exceptions.py\", line 71, in __call__\n    await self.app(scope, receive, sender)\n  File \"/usr/local/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py\", line 21, in __call__\n    raise e\n  File \"/usr/local/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py\", line 18, in __call__\n    await self.app(scope, receive, send)\n  File \"/usr/local/lib/python3.10/site-packages/starlette/routing.py\", line 656, in __call__\n    await route.handle(scope, receive, send)\n  File \"/usr/local/lib/python3.10/site-packages/starlette/routing.py\", line 259, in handle\n    await self.app(scope, receive, send)\n  File \"/usr/local/lib/python3.10/site-packages/starlette/routing.py\", line 61, in app\n    response = await func(request)\n  File \"/usr/local/lib/python3.10/site-packages/fastapi/routing.py\", line 227, in app\n    raw_response = await run_endpoint_function(\n  File \"/usr/local/lib/python3.10/site-packages/fastapi/routing.py\", line 162, in run_endpoint_function\n    return await run_in_threadpool(dependant.call, **values)\n  File \"/usr/local/lib/python3.10/site-packages/starlette/concurrency.py\", line 39, in run_in_threadpool\n    return await anyio.to_thread.run_sync(func, *args)\n  File \"/usr/local/lib/python3.10/site-packages/anyio/to_thread.py\", line 31, in run_sync\n    return await get_asynclib().run_sync_in_worker_thread(\n  File \"/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py\", line 937, in run_sync_in_worker_thread\n    return await future\n  File \"/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py\", line 867, in run\n    result = context.run(func, *args)\n  File \"/phovea/tdp_core/tdp_core/mol_img/img_api.py\", line 40, in draw_maximum_common_substructure_molecule\n    return draw(mcs)\n  File \"/phovea/tdp_core/tdp_core/mol_img/util/draw.py\", line 23, in inner\n    draw_inner(drawer, *args, **kwargs)\n  File \"/phovea/tdp_core/tdp_core/mol_img/util/draw.py\", line 34, in draw\n    drawer.DrawMolecule(structure, highlightAtoms=highlight_atoms, highlightBonds=None, highlightAtomColors=None, highlightBondColors=None)\nBoost.Python.ArgumentError: Python argument types in\n    MolDraw2D.DrawMolecule(MolDraw2DSVG, NoneType)\ndid not match C++ signature:\n    DrawMolecule(RDKit::MolDraw2D {lvalue} self, RDKit::ROMol mol, boost::python::api::object highlightAtoms, boost::python::api::object highlightBonds, boost::python::api::object highlightAtomColors=None, boost::python::api::object highlightBondColors=None, boost::python::api::object highlightAtomRadii=None, int confId=-1, std::string legend='')\n    DrawMolecule(RDKit::MolDraw2D {lvalue} self, RDKit::ROMol mol, boost::python::api::object highlightAtoms=None, boost::python::api::object highlightAtomColors=None, boost::python::api::object highlightAtomRadii=None, int confId=-1, std::string legend='')\n"}
```


@dvmarkusvogl Please have a look how you can fix these two bugs. Thanks.

**Comment by puehringer** [comment]
@dvmarkusvogl @thinkh Another thing I just noticed is the path, i.e. /api/image is not ideal, but maybe prefix it with /api/chem/... or /api/rdkit/...

**Comment by thinkh** [comment]
@dvmarkusvogl As discussed with Michael in https://github.com/datavisyn/tdp_core/issues/741#issuecomment-1184827576, we would need an additional API endpoint with the following spec:

**Payload**
* List of structures
* Query structure

**Result**
* List with `true` or `false` if the query matches the given structures

Can you please add such an endpoint? Do you need further information?

**Comment by dvmarkusvogl** [comment]
Fixed:
* Paths now start with /rdkit/ instead of /image/
* Change from URL params to Path parameters (default type for GET parameters, so you just have to delete the {structure} from the URL) for ALL endpoints that take molecules.
* Adapted tests
* Added the problematic MCS testdata. The actual call to rdkit returns None in some cases (giving the error you had, which was, that the C++ wrapper expected a molecule but got None) and sometimes a single QueryMolecule (drawn as ?) in other cases. This is an rdkit issue that also implements in the original implementation. I think we should start a separate issue for this if this input data would yield any useful output data (which it shouldn't according to my bioinf knowledge, considering `matchValences=True, ringMatchesRingOnly=True, completeRingsOnly=True`)

See API-Doc or tests on how to use it now, but tl;dr: instead of url params, use query params now.

Example for the problem molecule from @thinkh 
image

**Comment by thinkh** [comment]
@dvmarkusvogl Thanks for the fixes and changes. I'll integrate them in my PR and let you know if everything works.

**Review by thinkh — DISMISSED** [review]

**Review by puehringer — APPROVED** [review]
